Variable expansion outside queries

neelasha23 opened the pull request:
## Describe your changes

Added variable expansion support for the string type arguments in the following:
1. `%sqlcmd columns` 
2. `%sqlcmd explore` 
3. `%sqlcmd profile`
4. `%sqlcmd snippets`
5. `%sqlcmd tables`
6. `%sqlcmd test`
7. `%sql`
8. `%sqlplot`

Integer/ float arguments of histogram plot like `bins`, `binwidth`, etc are not supported.


## Issue number

Closes #699 

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--941.org.readthedocs.build/en/941/

<!-- readthedocs-preview jupysql end -->

neelasha23 commented:
I haven't added histogram specific arguments like `bins`, `binwidth` etc, and comparator arguments for testing like `greater`, `greater_or_equal` etc. Do we need to add these as well? @edublancas

edublancas commented:
@neelasha23 I think expanding the parameters when we read the arguments string would be better so we automatically add expansion to them. Otherwise, we'll have to remember to add this feature to every argument we add. If you look at the SQL magic implementation, there should be a part where we obtain the raw argument string e.g. `--a something --b another`; we could expand the values there

neelasha23 commented:
Addressed the comment @edublancas

edublancas commented:
@bryannho please review this, pay special attention to this [comment](https://github.com/ploomber/jupysql/pull/941#issuecomment-1874219746)

bryannho reviewed (commented):
Overall the variable expansion works well, but I have some questions. Of course the intention is to use variables to specify argument values like this:

```
%%sql --save {{foo}} --no-execute
SELECT 1
```

But in testing I found that you can also use them to specify the argument flags themselves, like this:

```
arg = "--save"
```

```
%%sql {{arg}} snippet --no-execute
SELECT 1
```

This works fine when the arguments are specified correctly. However, I could do something like

```
arg = "--not_an_arg"
```

and the statement will still run without throwing an error:

<img width="563" alt="image" src="https://github.com/ploomber/jupysql/assets/42661667/14252278-7f7d-4879-ae24-ca37e06eaf78">

@edublancas @neelasha23 Do we want to allow users to specify argument flags using variables in this way? I assume this wasn't the original intent. It could definitely be helpful but might also open up more issues in the future.

If so, I think we will need to call `expand_args()` before we parse/check for invalid arguments. We will also need to include this in the documentation and spend more time thinking about potential corner cases/bugs.

neelasha23 commented on doc/api/magic-sql.md:
```diff
@@ -359,3 +359,21 @@ LIMIT 3
+
+## Parameterizing arguments
+
+JupySQL supports variable expansion of arguments in the form of `{{variable}}`. This allows the user to specify arguments with placeholders that can be replaced by variables dynamically. The arguments supported are `section`, `close`, `alias`, `save` and `with`. 
```
Arguments list is still valid. But I've removed this sentence because everytime a new argument is added we need to remember to add to this list: https://github.com/ploomber/jupysql/pull/941/commits/a2c9266287d19c035e3e441580292fb0d72ec647

neelasha23 commented on doc/api/magic-plot.md:
```diff
@@ -293,3 +293,52 @@ You can also show the percentage on top of the pie using the `S`/`show-numbers`
+
+JupySQL supports variable expansion of arguments in the form of `{{variable}}`. This allows the user to specify arguments with placeholders that can be replaced by variables dynamically.
+
+Currently, `table`, `column`, `schema`, `with` and `orient` arguments of `%sqlplot` can be parametrized. Let's see an example:
```
allows parametrization of string arguments. So list is valid but removed this line completely because everytime a new argument is added we need to remember to add to this list: https://github.com/ploomber/jupysql/pull/941/commits/f26bccfb5676081d60845f0042bd551c5c784d62#diff-d2bac2aa48fbf4cb4aee34358d99fb9653fe443d5d3359ea2848dd36bdaa2f3bL301-L302

neelasha23 commented on src/sql/magic.py:
```diff
@@ -404,6 +403,8 @@ def interactive_execute_wrapper(**kwargs):
         user_ns = self.shell.user_ns.copy()
         user_ns.update(local_ns)
 
+        line = util.expand_args(line, user_ns)
```
This functions is changed now: https://github.com/ploomber/jupysql/pull/941/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R411-R412

neelasha23 commented:
> @edublancas @neelasha23 Do we want to allow users to specify argument flags using variables in this way? I assume this wasn't the original intent. It could definitely be helpful but might also open up more issues in the future.

I think for this PR we should only handle arguments and not flags. I have changed the approach. Previously I had tried a couple of approaches:
1. Regex based to extract `--arg {{value}}` patterns and then rendering. This works but I'm not very confident of this approach since false edge cases/ missed cases might come up.
2. Allow the `argparse` parser to parse the input as is and then substitute the args. So the `argparse.Namespace` would contain values like `save: {{snippet}}` and then `{{snippet}}` would be rendered by the actual snippet name. I feel this approach is less error prone and more deterministic, as now we are rendering only the argument values specifically, although it requires an additional step of iterating over the namespace. So I have reverted to this approach.

Here are the changes:


Regarding the error, now it will display error if we try to use the snippet:

<img width="848" alt="Screenshot 2024-01-23 at 4 59 36 PM" src="https://github.com/ploomber/jupysql/assets/29096135/3b4ce66a-a14d-4d85-8f70-7730fe264934">

bryannho reviewed (commented):
@neelasha23 Manual testing checks out. Overall your last solution was more streamlined but I can see why you needed to revert back to this approach in order to disallow parametrizing flags. I'll leave it up to @edublancas if he feels like this is too clunky. If we were to add a new command it would have to remember to include this logic:

```python
 args = parser.parse_args(others)

    if is_rendering_required(" ".join(others)):
        expand_args(args, user_ns)
```

Not sure if we even plan on adding new commands anytime soon though.

It would be helpful to add a warning to users when they try to parametrize a flag. Even though the flag isn't registered, currently the command still executes so users may be confused why it isn't working. So if the user tried this:

```python
arg = "--save"

%%sql {{arg}} snippet
SELECT 1
```

They should receive `UserWarning: Parametrization of argument flags is not allowed` or something like that.

edublancas commented:
> Allow the argparse parser to parse the input as is and then substitute the args. So the argparse.Namespace would contain values like save: {{snippet}} and then {{snippet}} would be rendered by the actual snippet name. I feel this approach is less error prone and more deterministic, as now we are rendering only the argument values specifically, although it requires an additional step of iterating over the namespace. So I have reverted to this approach.

I'm confused. this approach sounds like a much more robust alternative than adding code to each of the SQL commands, and you mentioned "I have reverted to this approach",

but then I looked at the code and there are a few snippets that look like this:

```python
   if is_rendering_required(" ".join(others)):
        expand_args(args, user_ns)
```

the objective is to have a single place for expanding the `{{variables}}` and I thought we had agreed to do that. but it seems like we still have an implementation that requires us adding the same snippet on each command?

neelasha23 commented:
> the objective is to have a single place for expanding the `{{variables}}` and I thought we had agreed to do that. but it seems like we still have an implementation that requires us adding the same snippet on each command?

What i understood from your previous comment, you mentioned that we we might forget to add every-time we add a new `argument` so we need a approach where we dont need to bother abut adding the arguments to a list. But in the current approach we dont need to maintain any list of that sort it automatically detects the "{{value}}" pattern inside argparse and replaces them. 
But we still need to call this util function at the different commands file right? like `sql`, `sqlplot`. and the different versions of `sqlcmd`. Also, `sqlcmd` is parsed differently than than the other two.

Now if we add any new command like `sqlcmd something` or `sqlsomething` only then we need to add this block of code otherwise not. Does this make sense?

@edublancas

edublancas commented:
@neelasha23 oh, you're right. I thought we could add a single piece of logic that would work for all commands (`%%sql`, `%sqlcmd` , etc). But then I realized that they are all implemented in different places as they are magic objects, so your approach makes sense.

just fix the conflicts, the integration tests and this should be good to go!

neelasha23 commented:
conflicts and tests are fixed now @edublancas
